### ISSUE TITLE
Consolidating documentation that is common across almost all BeeWare projects

### ISSUE THREAD

**Issue by dgelessus** [beeware/beeware#24]
There are a few documentation sections that appear in almost every BeeWare project's documentation, often with very similar or identical content. Examples of this include the how-to guides for contributing to the code and documentation. All BeeWare projects are hosted on GitHub, and most of them are Python-based with reStructuredText documentation. This means that these documentation sections almost always describe the same steps:

* For code contributions:
  * Clone the repo
  * Set up a virtual environment
  * Install test dependencies from a requirements.txt file
  * Run the test suite
* For documentation contributions
  * Install documentation build dependencies from a requirements.txt file
  * Run Sphinx to build the docs
  * View the docs in a browser

These instructions always describe the same processes, but the exact instructions often vary between projects - for example [Briefcase has Windows, Mac and Linux instructions that clone via HTTPS and use `venv`](https://briefcase.readthedocs.io/en/latest/how-to/contribute-code.html), whereas [Ouroboros has Unix-only instructions that clone via SSH and use `mkvirtualenv`](https://pybee-ouroboros.readthedocs.io/en/latest/internals/contributing.html).

In most cases these differences are accidental - often the instructions were initially copied from another BeeWare project, but afterwards they were modified independently in one project without updating the other projects. This is problematic - contributing fixes/improvements to these docs requires updating many independent projects, and there is no good reason why each BeeWare project should have different instructions for the same processes in the first place. To solve these accidental discrepancies, the common parts of the documentation could be moved into a central repository (such as this one).

However, there are a few complications with this. Sometimes there are good reasons why projects have differing documentation on the same topic. For example, [Briefcase uses PyTest for testing](https://briefcase.readthedocs.io/en/latest/how-to/contribute-code.html), whereas [rubicon-objc uses unittest](https://rubicon-objc.readthedocs.io/en/latest/how-to/contribute.html), and [VOC uses a custom test runner](https://voc.readthedocs.io/en/latest/how-to/development-env.html) - each of which needs to be set up and called differently. This kind of information has to be maintained per-project. It's also not clear how the central documentation would be shared across projects - should it be included into each project's documentation page (with template-style substitutions for project name, etc.), or should it be hosted in one place and linked from each project?

I would be interested in thoughts/opinions on whether it would be worthwile to consolidate parts of the documentation like this, or whether it would be too much work for too little benefit.

Note: I'm writing this from my viewpoint as a Rubicon-ObjC developer. I'm not an active contributor in any other BeeWare projects at the moment, and I'm less familiar with their development workflows and documentation, so I'd be interested in hearing what contributors to other BeeWare projects would think about this.

---

This is a follow-up to [this review discussion thread](https://github.com/beeware/rubicon-objc/pull/145#discussion_r362994061) on the rubicon-objc repo.

**Comment by freakboy3742** [comment]
Thanks for logging this, @dgelessus. As you've noted, there's lots of duplicated content between BeeWare's various projects, and we haven't always been consistent in updating every version of a document. Consolidating these sources of documentation (where possible) is definitely a desirable goal, if only for minimizing effort. 

However, the other factor I want to keep in mind is the UX of a developer coming to the project. 

Even assuming that the clone/test/document instructions were *identical*, moving the documentation to a unified location would mean two things:

1) The instructions would be "generic" rather than specific. You can't give a copy-and-pasteable command line to clone a repo that will work for all repos, so that means you have to give "git clone <insert URL here>"-type instructions. It's not a big thing for an experienced developer, but the experience we've had in sprint situations is that many enthusiastic contributors don't necessarily have a lot of git experience, and every hurdle is one more thing 

2) Although the "big picture" is that there is a single BeeWare project with common procedures, individual contribuors to Toga/Rubicon/etc won't necessarily see that - they're looking at the individual library they're contributing to. If I were looking to contribute to Project X, I'd expect to find a contribution guide in Project X's documentation; if I got bounced to some other project's documentation, it could potentially be disorienting. It's at least one more click, plus some explanatory text about why I need to read about "BeeWare" when I want to contribute to "Toga". 

Both of these are potentially manageable - but we need to keep them in mind as use cases for the consumption of these docs.

**Comment by dgelessus** [comment]
> However, the other factor I want to keep in mind is the UX of a developer coming to the project.

I completely agree with both of your points here, and this is perhaps my biggest worry with this plan.  Of all the documentation sections, the guides for new contributors have perhaps the most overlap across projects, but they are also the ones that benefit most from being written specifically for each project.

An ideal solution IMHO would be a template-based system: there would be a single shared "getting started" snippet, with placeholders for project name, Git repo URL, etc., which could be included in each project's docs as appropriate. This would be a purely internal change - if done properly, the end result seen by readers would be identical to what we have currently (except more up-to-date). I don't know if something like this can be realistically implemented with Sphinx though.

It might also be worth considering a different solution for more advanced common documentation, such as "how to release a new version of <project> to PyPI" (which is the topic where this discussion initially started). Someone in charge of handling project releases will probably already know about the BeeWare project and its relation to the specific project in question, so a general page under the common BeeWare docs will be less confusing to them compared to someone new to BeeWare.

**Comment by dgelessus** [comment]
I came across this issue again recently and thought about it some more.

Looking at the "getting started"/"how to contribute" guides for example, I don't think they can be consolidated in the way I originally thought, by creating templates in a separate repo that are automatically included in each project with a few variable substitutions. The guides in each project are simply too different, and for good reason - they usually contain some suggestions for where to start contributing, or instructions and warnings that apply only to that specific project. Although it might be *possible* to reformat all of these guides to fit into a common template, it simply isn't worth the effort if the guides are so different that every project has to override almost every part of the template anyway.

And that's not even considering the technical complexity needed to implement this. reStructuredText supports some template-like features using [`include`](https://docutils.sourceforge.io/docs/ref/rst/directives.html#include) directives and [substitutions](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#substitution-definitions), but they are limited to simple literal text substitutions. There's also no standard mechanism for "importing" files from another project (à la pip), so we'd need to roll our own using Git submodules, and those are *not* beginner-friendly. In fact, *any* kind of complex templating mechanism won't be beginner-friendly, and especially the beginner documentation needs to be easy for beginners to contribute to.

---

I still think it's a good idea to have a common documentation template, just not an *automatic* one. Instead I think it would be better to have a manual "copy-paste template" repo, i. e. a set of templates for common documentation pages, which can be copied into projects and customized as needed. We could also make the template repo compatible with Cookiecutter, to allow easily generating a basic set of documentation pages for a project that doesn't have any docs yet. (This would be an improvement over how it's been done in the past - if a project doesn't have docs yet, copy the common doc pages from another project and manually adjust the project name everywhere.)

Compared to an automatic template system, the advantage of a manual template like this is that the files in the project repos are still regular plain reST files, so there is no overhead in terms of technical complexity. To work on the docs, you don't have to understand how the template works (or that it even exists), you can simply edit the project's reST files as you would normally. This also allows much more flexibility in customizing the documentation for each individual project - if part of the template doesn't fit the project, or you need to insert extra information somewhere, you're free to do that without being constrained by what parameters the template has.

The obvious disadvantage of this approach is that it's entirely manual. There's nothing that keeps the template in sync with the projects that use it - when the template repo is updated, the same change needs to be performed manually in each project based on the template. Similarly, when part of the template is used in a project, and then the part from the template is improved in the project, that improvement needs to be manually submitted back to the template repo (and then ported from there to all other projects that use the template).

This means that a non-automatic template like this won't fully fix the problem of inconsistent docs between projects - each project's docs still need to be maintained and updated manually, even if they were originally based on the template. But even if it's entirely manual, the advantage of a template repo like this is that it serves as a single reference point for the common documentation pages. When starting the documentation for a project you can use the template as a base, and when updating outdated docs in a project you can use parts of the template instead of manually rewriting the outdated parts. Also, having a single reference point helps answer questions like whether to use `venv`/`virtualenv`/`mkvirtualenv`, `pip` vs. `python -m pip`, `git clone` over HTTPS vs. SSH, etc.

---

With that long explanation out of the way... Compared to my previous ideas on this topic, this one should be a bit more realistic. I think I'll try to create a few template doc pages based on some common pages from our existing project docs (like the getting started guides), to see how well the general idea works. (Can't promise when I'll get around to it, but I'll post here if I make any notable progress.)

**Comment by freakboy3742** [comment]
beeware/.github#228 proposes an approach that would allow us to use a template repository for some file content such as README content... but in discussion on Discord, @HalfWhitt made the suggestion that the same approach could be used to coordinate common *documentation* as well.
